Consistent formatting of all columns in Excel exports

markbastian opened the pull request:
This PR has a set of changes that should provide consistent formatting across exports of Excel, CSV, and JSON files, including formatting applied globally, as metadata, and as visualization-specific settings.

Changes include:
- Consolidation of settings merge logic into `metabase.query-processor.streaming.common` as `viz-settings-for-col`. This is now used for `xlsx`, `csv`, and `json` exports.
  - The `xlsx` still needs to take these settings and produce cell formatters where the `csv` and `json` branches need to produce string formatters. Either way, this is a good incremental step towards a pipeline of all the settings -> unified settings per column -> format -> product of format (cell style or formatter).
  - This also provides a more functional style in which we reach for global settings far less and rely on what is passed in via the unified settings (which get global settings from the `metabase.query-processor.middleware.visualization-settings` qp middleware).
- Modify `metabase.query-processor.streaming.xlsx` to use the same strategy as csv and json exports of computing styles for visible columns only up front then reusing these styles
- Removing trailing zeroes for percentages in CSV and JSON files prior to formatting to conform to the FE. While this gets us close to consistent behavior, we should maybe revisit the idea of default stripping of trailing zeroes.
- Adds consistent seed values of `nil` to the `volatile!` fields used in stream processing for column formatting.
  - Note that this ns still has some potential to be streamlined to remove code that is close to code in the csv/json path but this PR is meant to be a start, not to completely refactor the whole ns.
- Unit tests
  - All existing tests pass. Note that `metabase.query-processor.streaming.xlsx-test/format-string` has been modified to compute a unified settings map to better follow what happens in practice with the qp middleware.
  - Additional smaller and large-scale tests were added to `metabase.api.card-test` to test end to end xlsx exports.

Fixes #14393
Closes #17753

markbastian committed:
Removing Time portion of Timestamp when formatted as Date in Excel

This is a starting changeset for fixing the issue "Timestamp is always present in dates when exporting to Excel in the embedded application. #14393", which demonstrates a failure mode in applying column-formatting to Excel downloads. This commit should fix the underlying issue as well as update formatting for remaining CSV file download discrepancies.

The goal of this commit is to get us to a baseline state for some small refactorings to combine our formatting paths for CSV, JSON, and Excel exports.

markbastian committed:
Updating common logic and tests to reflect how global settings are passed in and applied to format-temporal-str

markbastian committed:
Updating `number-format-strings` to not need semantic type.

markbastian committed:
Adding tests for new Excel-export logic.

adam-james-v commented on src/metabase/formatter.clj:
```diff
@@ -112,8 +112,10 @@
-              decimals-in-value (digits-after-decimal (if percent? (* 100 scaled-value) scaled-value))
+        (let [scaled-value      (cond-> (* value (or scale 1))
+                                  percent?
+                                  (* 100))
```
Yea, this is a nice small change that makes the code more sensible :)

markbastian replied:
It also gives consistency with the FE by trimming trailing zeros for percentages. I'm still not sure I like that behavior, but it is consistent.

markbastian committed:
Fixing cases for currency where we must specify that `::mb.viz/currency-in-header` is false.

adam-james-v commented on src/metabase/formatter.clj:
```diff
@@ -127,22 +129,25 @@
+                              (cond-> (.format fmtr scaled-value)
+                                (and (not currency?) (not decimals))
+                                (strip-trailing-zeroes decimal)
+                                percent?    (str "%"))
```
I can't quite follow why this `(str "5")` was moved from the `fmt-str` in the let binding? 
Whatever the logic is for moving it, should it also be applied to the `scientific?` pred that is still in `fmt-str`?

markbastian replied:
This was done so that trailing zeros could be trimmed from the string. This is to be consistent with the FE.

adam-james-v replied:
Ah, of course! Thanks

adam-james-v commented on src/metabase/query_processor/streaming/common.clj:
```diff
@@ -69,7 +70,7 @@
   for a single column."
   [format-settings global-settings-key]
-  (let [global-settings (global-settings-key (public-settings/custom-formatting))
+  (let [global-settings (get (public-settings/custom-formatting) global-settings-key)
```
This is just curiosity: why is `(get {} :key)` preferred here over `(:key {})`? Just style pref? Something more consequential?

markbastian replied:
Normally I just use the key as a function, but since the key is dereferenced value I felt that this gives some level of self-documentation -- it's obvious this is a lookup and not a ns-level function. I don't have a strong opinion on reverting this line, though.

adam-james-v replied:
No, I think keep it. I was really just curious what the thought was. Thanks for the explanation.

adam-james-v commented on src/metabase/query_processor/streaming/common.clj:
```diff
@@ -112,3 +113,64 @@
         (str column-title " (" (currency-identifier merged-settings) ")")
         column-title))))
+
+(defn normalize-keys
```
We have `metabase.util/normalize-map` and a few various key normalization functions in the util namespace. Should we consider using a util fn instead? 

I'm happy with this solution too, it's small and straightforward. Just wanted to mention I saw that util fn in case you weren't aware.

markbastian replied:
I'm actually thinking maybe we should drop this entirely and force `metabase.formatter.datetime` to use fully-qualified keys.

adam-james-v replied:
That's not a bad idea, actually.

markbastian committed:
Fixing formatting tests

markbastian committed:
Updating unit test to reflect a global abbreviated dates preference.

adam-james-v commented on src/metabase/query_processor/streaming/common.clj:
```diff
@@ -112,3 +113,64 @@
+                              ::mb.viz/column-settings
+                              ;; update the keys so that they will have only the :field-id or :column-name
+                              ;; and not have any metadata. Since we don't know the metadata, we can never
+                              ;; match a key with metadata, even if we do have the correct name or id
```
This part is debatable right? Like, in a Slack thread I thought it was discussed that column name or field-id is not technically sufficient to disambiguate columns in all cases. 

This assumption (dropping metadata) has been the behaviour already I'm pretty sure, so this PR shouldn't try to fix it, but this is something we should revisit later maybe? Please disregard this if I'm misremembering.

adam-james-v commented on test/metabase/api/card_test.clj:
```diff
@@ -1934,6 +1935,172 @@
+                    ["Jan 1, 2023, 12:34 PM" "2023-1-1" "1-1-2023, 12:34:56.000" "Jan 1, 2023" "12:34 PM" "2,345.30" "[$$]2,345.30" "[$USD] 3456.30" "2,931.30 US dollars" "1.71806E+4" "8.02%" "0.000%"]
+                    ["Jan 1, 2023, 12:34 PM" "2023-1-1" "1-1-2023, 12:34:56.000" "Jan 1, 2023" "12:34 PM" "3,456.00" "[$$]3,456.00" "[$USD] 2300.00" "2,250.00 US dollars" "12.7181E+4" "95.40%" "11.580%"]]
+                   (parse-xlsx-results-to-strings
+                     (mt/user-http-request :rasta :post 200 (format "card/%d/query/xlsx" (u/the-id card))))))))))))
```
I'm glad there's a test for all of the formatting. If something breaks, will the test failure be readable? I suppose the diff will show fairly quickly which formatter is not working as expected?

The alternative I can think of is tons of tiny tests, which I'm not sure what's better or worse?

adam-james-v reviewed (approved):
Any comments/questions I left are more curiosity or minor considerations, none of which I think should block this PR, so I'm approving it.

markbastian committed:
Adding column type defaults for certain test cases.

markbastian committed:
Updating logic and tests in xlsx exports

Logic in `metabase.query-processor.streaming.common` further unifies the idea of merging all settings first, then making decisions based on that. It also makes the downstream code more functional, as that code doesn't repeatedly reach for global settings.

`metabase.query-processor.streaming.xlsx-test` has been updated to reflect this change. In particular, the helper function `format-string` now builds the right input settings rather than passing settings through to be accumulated later.

markbastian committed:
Fixing two more tests.

markbastian committed:
Fixing remaining unit tests

- Update `metabase.api.card-test` with multiple test cases for different formats.
- Update `export-format-test` to lock temporary settings for consistency.
- `default-number-format?` to `unformatted-number?` and include "currency" as a potential unformatted type.

markbastian committed:
Consistent application of global viz settings

This PR adds a function that will inject the global viz settings into `viz-settings-for-col` if not provided by the query processor. This also simplifies the `xslx-test` ns.

markbastian committed:
Treating :default unit as no unit.

GeorgeVragalev commented:
when is this going to be deployed and available?
as from what I can see now the csv download from metabase UI for questions doesn't match the table columns that were made to have custom names
